pkg/authn/kubernetes/keychain.go

```diff
@@ -0,0 +1,161 @@
+	// will run (scoped to Namespace).  If empty, "default" is assumed.
+	ServiceAccountName string
+	// ImagePullSecrets holds the names of the Kubernetes secrets (scoped to
+	// Namespace) containing credential data to use for the image pull.
+	ImagePullSecrets []string
+}
+
+// New returns a new authn.Keychain suitable for resolving image references as
+// scoped by the provided Options.  It speaks to Kubernetes through the provided
+// client interface.
+func New(ctx context.Context, client kubernetes.Interface, opt Options) (authn.Keychain, error) {
+	if opt.Namespace == "" {
+		opt.Namespace = "default"
+	}
+	if opt.ServiceAccountName == "" {
+		opt.ServiceAccountName = "default"
+	}
+
+	// Implement a Kubernetes-style authentication keychain.
+	// This needs to support roughly the following kinds of authentication:
+	//  1) The explicit authentication from imagePullSecrets on Pod
+	//  2) The semi-implicit authentication where imagePullSecrets are on the
+	//    Pod's service account.
+
+	// First, fetch all of the explicitly declared pull secrets
+	var pullSecrets []corev1.Secret
+	for _, name := range opt.ImagePullSecrets {
+		ps, err := client.CoreV1().Secrets(opt.Namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		pullSecrets = append(pullSecrets, *ps)
+	}
+
+	// Second, fetch all of the pull secrets attached to our service account.
+	sa, err := client.CoreV1().ServiceAccounts(opt.Namespace).Get(ctx, opt.ServiceAccountName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, localObj := range sa.ImagePullSecrets {
+		ps, err := client.CoreV1().Secrets(opt.Namespace).Get(ctx, localObj.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		pullSecrets = append(pullSecrets, *ps)
+	}
+
+	return NewFromPullSecrets(ctx, pullSecrets)
+}
+
+// NewInCluster returns a new authn.Keychain suitable for resolving image references as
+// scoped by the provided Options, constructing a kubernetes.Interface based on in-cluster
+// authentication.
+func NewInCluster(ctx context.Context, opt Options) (authn.Keychain, error) {
+	clusterConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		return nil, err
+	}
+	return New(ctx, client, opt)
+}
+
+// NewFromPullSecrets returns a new authn.Keychain suitable for resolving image references as
+// scoped by the pull secrets.
+func NewFromPullSecrets(ctx context.Context, secrets []corev1.Secret) (authn.Keychain, error) {
+	m := map[string]authn.AuthConfig{}
+	for _, secret := range secrets {
+		if b, exists := secret.Data[corev1.DockerConfigJsonKey]; secret.Type == corev1.SecretTypeDockerConfigJson && exists && len(b) > 0 {
+			var cfg config
+			if err := json.Unmarshal(b, &cfg); err != nil {
+				return nil, err
+			}
+			for k, v := range cfg.Auths {
+				m[k] = v
```

Comment:
This (and line 119 below) means that later-specified secrets with matching repo/registry keys will overwrite previous ones. `New` fetches explicitly listed `imagePullSecrets` before searching for those attached to the SA, so those later ones, if there's a conflict, will win.

We have some options:
- don't overwrite -- first matching auth wins
- fetch implicit SA secrets before explicitly listed secrets

In any case, we probably need to spell it out better and guard it with tests.

Reply:
Just to chime in here (I meant to before).  If this matches K8s' behavior, then it SGTM.  It's unfortunate we don't track that anymore, but it felt like that was only a matter of time anyways.

Reply:
Well, it turns out the previous behavior was first-one-wins, which I verified with a test on `main` that I ported to this PR.